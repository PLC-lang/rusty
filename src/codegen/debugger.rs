
use inkwell::debug_info::{DebugInfoBuilder,  DICompileUnit, DIType, DIScope, DISubprogram, AsDIScope};
use inkwell::context::Context;
use inkwell::builder::Builder;

use crate::ast::*;
use crate::index::DataTypeInformation;

pub struct DebugManager<'ctx> {

    context: Option<&'ctx Context>,
    di_builder: Option<DebugInfoBuilder<'ctx>>,
    compilation_unit: Option<DICompileUnit<'ctx>>,
    scope:  Option<DIScope<'ctx>>,
}

impl <'ctx> DebugManager<'ctx> {
    pub fn create_inactive() -> DebugManager<'ctx> {
        DebugManager::new(None,None, None)
    }

    pub fn activate(&mut self, context: &'ctx Context, di_builder : DebugInfoBuilder<'ctx>, compilation_unit: DICompileUnit<'ctx>) {
        self.context = Some(context);
        self.di_builder = Some(di_builder);
        self.compilation_unit = Some(compilation_unit);
    }
    
    pub fn new(context: Option<&'ctx Context>,  di_builder : Option<DebugInfoBuilder<'ctx>>, compilation_unit: Option<DICompileUnit<'ctx>>) -> DebugManager<'ctx> {
        DebugManager {
            context,
            di_builder,
            compilation_unit,
            scope: None,
        }
    }

    pub fn finalize(&self) {
        if let Some(di_builder) =  &self.di_builder {
            di_builder.finalize();       
        }
    }

    pub fn create_subprogram(&mut self, members : &[(&str, DataTypeInformation<'ctx>)], function : &POU, new_lines : &[usize]) -> Option<(DIType<'ctx>, DISubprogram<'ctx>)>{
        //Exit if dummy
        if self.di_builder.is_none() {
            return None;
        }

        let di_builder = self.di_builder.as_ref().unwrap();
        let compilation_unit = self.compilation_unit.as_ref().unwrap();

        let line_number = get_line_of(new_lines, &function.location.start) as u32 +1 ; //TODO: Line numbers are off by one
        //
        //Create and get a struct type
        let struct_type = self.get_struct_type(function.name.as_str(), line_number, members).unwrap();

        //Add function debug information
        let scope = self.get_current_scope();
        let subroutine_type = di_builder.create_subroutine_type(compilation_unit.get_file(), 
        None,  &[struct_type] , inkwell::debug_info::DIFlagsConstants::PUBLIC, );
        let result = di_builder.create_function(
            scope, 
            function.name.as_str(), 
            None, 
            compilation_unit.get_file(), 
            line_number, 
            subroutine_type, 
            true, 
            true, 
            line_number, 
            inkwell::debug_info::DIFlagsConstants::PUBLIC, 
            false
        );
        self.scope = Some(result.as_debug_info_scope());
        Some((struct_type, result))
    }

    fn get_current_scope(&self) -> DIScope<'ctx> {
        self.scope.or(Some(self.compilation_unit.as_ref().unwrap().get_file().as_debug_info_scope())).unwrap()
    }

    pub fn get_basic_type(&self, name : &str, datatype: &DataTypeInformation) -> Option<DIType<'ctx>> {
         //Exit if dummy
         if self.di_builder.is_none() {
            return None;
        }

        let di_builder = self.di_builder.as_ref().unwrap();
        //TODO : Cache
        match *datatype {
            DataTypeInformation::Integer {size, ..} => {
                di_builder.create_basic_type(
                    name, 
                    size as u64, 
                    0x00, 
                    inkwell::debug_info::DIFlagsConstants::PUBLIC
                ).map(|it| it.as_type()).ok()
            },
            _ => None,
        }   
    }
/*
    pub fn get_instance_variable(&self, name : &str, debug_type : DIType<'ctx>, line : u32,global_value : PointerValue,block : BasicBlock, debug_location : DILocation) {
        //Exit if dummy
        if self.di_builder.is_none() {
            return;
        }

        let context = self.context.unwrap();
        let di_builder = self.di_builder.as_ref().unwrap();
        let compilation_unit = self.compilation_unit.as_ref().unwrap();
        
        let instance_variable = Some(di_builder.create_parameter_variable(
            self.get_current_scope(), 
            name, 
            0, 
            compilation_unit.get_file(), 
            line, 
            debug_type, 
            false, 
            inkwell::debug_info::DIFlagsConstants::PUBLIC,
        ));

        let debug_loc = di_builder.create_debug_location(context, line, 0, self.get_current_scope(), None);
        di_builder.insert_declare_at_end(
            global_value,
             instance_variable, 
             Some(di_builder.create_expression(Vec::new())), 
             debug_loc, 
             block
        );
    }
*/
    /* fn get_variables_debug_information(
        &self,
        debugger: &mut DebugManager<'ctx>,
        variables: &Vec<Variable>,
    ) -> Vec<(String, Option<DIType<'ctx>>)>
    {
        let mut types: Vec<(String, Option<DIType<'ctx>>)> = Vec::new();
        for variable in variables {
            let (_, debug_type)  = self.get_debug_type(debugger, &variable.data_type).unwrap();
            types.push((variable.name.clone(), debug_type));
        }
        types
    }

    fn get_debug_type(&self, name: &str, data_type: &DataTypeInformation<'ctx>) -> Option<(BasicTypeEnum<'ctx>, Option<DIType<'ctx>>)> {
        Some(data_type.get_type(), self.get_basic_type(name, data_type))
    }
*/
    pub fn get_struct_type(&self, name : &str, line : u32, members : &[(&str, DataTypeInformation<'ctx>)]) -> Option<DIType<'ctx>> {
        //Exit if dummy
        if self.di_builder.is_none() {
            return None;
        }

        let di_builder = self.di_builder.as_ref().unwrap();
        let compilation_unit = self.compilation_unit.as_ref().unwrap();


        let mut elements = Vec::new();
        let mut size = 0;
        for (name, datatype) in members {
            size = size + datatype.get_size();
            if let Some(member) = self.get_basic_type(name, datatype) {
                elements.push(member);
            }
        }

        let result = di_builder.create_struct_type(
            self.get_current_scope(), 
            name, 
            compilation_unit.get_file(), 
            line, 
            size as u64, 
            8, 
            inkwell::debug_info::DIFlagsConstants::PUBLIC, 
            None, 
            elements.as_slice(), 
            0, 
            None, 
            format!("{}_debug_interface", name.to_string()).as_str(),
        );

        di_builder.create_typedef(result.as_type(), name, compilation_unit.get_file(), line
            , self.get_current_scope(), 8);


        let mut current_offset = 0;
        for (i, (name, datatype)) in members.iter().enumerate() {
            di_builder.create_member_type(result.as_debug_info_scope(), name, compilation_unit.get_file(), line, (datatype.get_size() * 8) as u64, 8, current_offset, 
                inkwell::debug_info::DIFlagsConstants::PUBLIC, 
                elements[i],
            );
            current_offset = current_offset + (datatype.get_size() * 8) as u64; //TODO : Add align magic
        }

        Some(result.as_type())
    }

    pub fn append_breakpoint_location(&mut self, builder : &Builder, line : u32) {
        //Exit if dummy
        if self.di_builder.is_none() {
            return;
        }
        let context = self.context.unwrap();
        let di_builder = self.di_builder.as_ref().unwrap();
        
        let parent_scope = self.get_current_scope();
        //let compilation_unit = self.compilation_unit.as_ref().unwrap();
        //let lexical_block = di_builder.create_lexical_block(parent_scope, compilation_unit.get_file(), line, 0);
        
        let result = di_builder.create_debug_location(context, line, 0, parent_scope, None);
        builder.set_current_debug_location(context, result)

    }

    pub fn remove_scope(&mut self) {
        self.scope = None;
    }
}
