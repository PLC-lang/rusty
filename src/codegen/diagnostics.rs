use std::{backtrace, fmt::Display};

use inkwell::builder::BuilderError;
use plc_diagnostics::diagnostics::Diagnostic;
use plc_source::source_location::SourceLocation;
use thiserror::Error;

#[derive(Error, Debug)]
pub struct CodegenDiagnostic {
    message: String,
    source: BuilderError,
    location: SourceLocation,
}

impl Display for CodegenDiagnostic {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "Codegen error: {}. Source: {} at {}", &self.message, &self.source, &self.location)
    }
}

impl From<BuilderError> for CodegenDiagnostic {
    fn from(source: BuilderError) -> Self {
        // println!("Backtrace: {} ", backtrace::Backtrace::force_capture());
        CodegenDiagnostic { message: source.to_string(), source, location: SourceLocation::undefined() }
    }
}

impl CodegenDiagnostic {
    pub(crate) fn with_location(self, location: SourceLocation) -> CodegenDiagnostic {
        CodegenDiagnostic { location, ..self }
    }
}

impl From<CodegenDiagnostic> for Diagnostic {
    fn from(value: CodegenDiagnostic) -> Self {
        Diagnostic::codegen_error(&value.message, value.location.clone()).with_internal_error(value.into())
    }
}
