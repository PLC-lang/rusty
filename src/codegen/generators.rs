pub mod data_type_generator;
pub mod date_time_util;
pub mod expression_generator;
pub mod llvm;
pub mod pou_generator;
pub mod statement_generator;
pub mod variable_generator;

// see: https://llvm.org/doxygen/NVPTXBaseInfo_8h_source.html
pub const ADDRESS_SPACE_GENERIC: u16 = 0;
pub const ADDRESS_SPACE_GLOBAL: u16 = 1;
pub const ADDRESS_SPACE_CONST: u16 = 4;
