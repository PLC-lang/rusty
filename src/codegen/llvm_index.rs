// Copyright (c) 2020 Ghaith Hachem and Mathias Rieder
use crate::{ast::SourceRange, compile_error::CompileError};
use inkwell::types::BasicTypeEnum;
use inkwell::values::{BasicValueEnum, FunctionValue, GlobalValue, PointerValue};
use std::collections::HashMap;

/// Index view containing declared values for the current context
/// Parent Index is the a fallback lookup index for values not declared locally
#[derive(Debug, Clone)]
pub struct LlvmTypedIndex<'ink> {
    parent_index: Option<&'ink LlvmTypedIndex<'ink>>,
    type_associations: HashMap<String, BasicTypeEnum<'ink>>,
    initial_value_associations: HashMap<String, BasicValueEnum<'ink>>,
    loaded_variable_associations: HashMap<String, PointerValue<'ink>>,
    implementations: HashMap<String, FunctionValue<'ink>>,
    constants: HashMap<String, BasicValueEnum<'ink>>,
}

impl<'ink> LlvmTypedIndex<'ink> {
    pub fn new() -> LlvmTypedIndex<'ink> {
        LlvmTypedIndex {
            parent_index: None,
            type_associations: HashMap::new(),
            initial_value_associations: HashMap::new(),
            loaded_variable_associations: HashMap::new(),
            implementations: HashMap::new(),
            constants: HashMap::new(),
        }
    }

    pub fn create_child(parent: &'ink LlvmTypedIndex<'ink>) -> LlvmTypedIndex<'ink> {
        LlvmTypedIndex {
            parent_index: Some(parent),
            type_associations: HashMap::new(),
            initial_value_associations: HashMap::new(),
            loaded_variable_associations: HashMap::new(),
            implementations: HashMap::new(),
            constants: HashMap::new(),
        }
    }

    pub fn merge(&mut self, mut other: LlvmTypedIndex<'ink>) {
        for (name, assocication) in other.type_associations.drain() {
            self.type_associations.insert(name, assocication);
        }
        for (name, assocication) in other.initial_value_associations.drain() {
            self.initial_value_associations.insert(name, assocication);
        }
        for (name, assocication) in other.initial_value_associations.drain() {
            self.initial_value_associations.insert(name, assocication);
        }
        for (name, assocication) in other.loaded_variable_associations.drain() {
            self.loaded_variable_associations.insert(name, assocication);
        }
        for (name, implementation) in other.implementations.drain() {
            self.implementations.insert(name, implementation);
        }
        self.constants.extend(other.constants);
    }

    pub fn associate_type(
        &mut self,
        type_name: &str,
        target_type: BasicTypeEnum<'ink>,
    ) -> Result<(), CompileError> {
        self.type_associations
            .insert(type_name.to_lowercase(), target_type);
        Ok(())
    }

    pub fn associate_initial_value(
        &mut self,
        type_name: &str,
        initial_value: BasicValueEnum<'ink>,
    ) -> Result<(), CompileError> {
        self.initial_value_associations
            .insert(type_name.to_lowercase(), initial_value);
        Ok(())
    }

    pub fn associate_loaded_local_variable(
        &mut self,
        container_name: &str,
        variable_name: &str,
        target_value: PointerValue<'ink>,
    ) -> Result<(), CompileError> {
        let qualified_name = format!("{}.{}", container_name, variable_name);
        self.loaded_variable_associations
            .insert(qualified_name.to_lowercase(), target_value);
        Ok(())
    }

    pub fn find_associated_type(&self, type_name: &str) -> Option<BasicTypeEnum<'ink>> {
        self.type_associations
            .get(&type_name.to_lowercase())
            .copied()
            .or_else(|| {
                self.parent_index
                    .map(|it| it.find_associated_type(type_name))
                    .flatten()
            })
    }

    pub fn get_associated_type(
        &self,
        type_name: &str,
    ) -> Result<BasicTypeEnum<'ink>, CompileError> {
        self.find_associated_type(type_name)
            .ok_or_else(|| CompileError::unknown_type(type_name, SourceRange::undefined()))
    }

    pub fn find_associated_initial_value(&self, type_name: &str) -> Option<BasicValueEnum<'ink>> {
        self.initial_value_associations
            .get(&type_name.to_lowercase())
            .copied()
            .or_else(|| {
                self.parent_index
                    .map(|it| it.find_associated_initial_value(type_name))
                    .flatten()
            })
    }

    pub fn associate_global(
        &mut self,
        variable_name: &str,
        global_variable: GlobalValue<'ink>,
    ) -> Result<(), CompileError> {
        self.initial_value_associations.insert(
            variable_name.to_lowercase(),
            global_variable.as_pointer_value().into(),
        );
        Ok(())
    }

    pub fn associate_implementation(
        &mut self,
        callable_name: &str,
        function_value: FunctionValue<'ink>,
    ) -> Result<(), CompileError> {
        self.implementations
            .insert(callable_name.to_lowercase(), function_value);
        Ok(())
    }

    pub fn find_associated_implementation(
        &self,
        callable_name: &str,
    ) -> Option<FunctionValue<'ink>> {
        self.implementations
            .get(&callable_name.to_lowercase())
            .copied()
            .or_else(|| {
                self.parent_index
                    .map(|it| it.find_associated_implementation(callable_name))
                    .flatten()
            })
    }

    pub fn find_associated_variable_value(
        &self,
        qualified_name: &str,
    ) -> Option<BasicValueEnum<'ink>> {
        self.initial_value_associations
            .get(&qualified_name.to_lowercase())
            .copied()
            .or_else(|| {
                self.parent_index
                    .map(|it| it.find_associated_variable_value(qualified_name))
                    .flatten()
            })
    }

    pub fn find_loaded_associated_variable_value(
        &self,
        qualified_name: &str,
    ) -> Option<PointerValue<'ink>> {
        let result = self
            .loaded_variable_associations
            .get(&qualified_name.to_lowercase())
            .copied()
            .or_else(|| {
                self.parent_index
                    .map(|it| it.find_loaded_associated_variable_value(qualified_name))
                    .flatten()
            });

        //If nothing got associated, see if we have a global we could reuse
        result.or_else(|| {
            self.find_associated_variable_value(qualified_name)
                .filter(|it| it.is_pointer_value())
                .map(BasicValueEnum::into_pointer_value)
        })
    }

    pub fn associate_constant(
        &mut self,
        qualified_name: &str,
        basic_value_enum: BasicValueEnum<'ink>,
    ) {
        self.constants
            .insert(qualified_name.into(), basic_value_enum);
    }

    pub fn find_constant_value(&self, qualified_name: &str) -> Option<BasicValueEnum<'ink>> {
        self.constants.get(qualified_name).copied()
    }
}
