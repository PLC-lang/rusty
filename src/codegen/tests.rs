// Copyright (c) 2020 Ghaith Hachem and Mathias Rieder
mod code_gen_tests;
mod codegen_error_messages_tests;
mod compare_instructions_tests;
mod constants_tests;
mod debug_tests;
mod directaccess_test;
mod expression_tests;
mod function_tests;
mod generics_test;
mod initialization_test;
mod parameters_tests;
mod statement_codegen_test;
mod string_tests;
#[cfg(feature = "verify")]
mod switch_case_tests;
mod typesystem_test;
