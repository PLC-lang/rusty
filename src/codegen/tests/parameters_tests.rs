use crate::test_utils::tests::codegen;

#[test]
fn function_all_parameters_assigned() {
    // GIVEN
    let result = codegen(
        "
		FUNCTION foo : DINT
		VAR_INPUT
			input1 : DINT;
		END_VAR
		VAR_OUTPUT
			output1 : DINT;
		END_VAR
		VAR_IN_OUT
			inout1 : DINT;
		END_VAR
		END_FUNCTION

		PROGRAM main
		VAR
			var1, var2, var3 : DINT;
		END_VAR
			foo(var1, var2, var3);
			foo(input1 := var1, output1 => var2, inout1 := var3);
			foo(inout1 := var3, input1 := var1, output1 => var2);
		END_PROGRAM
		",
    );
    // THEN
    insta::assert_snapshot!(result);
}

#[test]
fn function_empty_input_assignment() {
    // GIVEN
    let result = codegen(
        "
		FUNCTION foo : DINT
		VAR_INPUT
			input1 : DINT;
		END_VAR
		VAR_OUTPUT
			output1 : DINT;
		END_VAR
		VAR_IN_OUT
			inout1 : DINT;
		END_VAR
		END_FUNCTION

		PROGRAM main
		VAR
			var1, var2, var3 : DINT;
		END_VAR
			foo(input1 := , output1 => var2, inout1 := var3);
		END_PROGRAM
		",
    );
    // THEN
    insta::assert_snapshot!(result);
}

#[test]
fn function_empty_output_assignment() {
    // GIVEN
    let result = codegen(
        "
		FUNCTION foo : DINT
		VAR_INPUT
			input1 : DINT;
		END_VAR
		VAR_OUTPUT
			output1 : DINT;
		END_VAR
		VAR_IN_OUT
			inout1 : DINT;
		END_VAR
		END_FUNCTION

		PROGRAM main
		VAR
			var1, var2, var3 : DINT;
		END_VAR
			foo(input1 := var1, output1 => , inout1 := var3);
		END_PROGRAM
		",
    );
    // THEN
    insta::assert_snapshot!(result);
}

#[test]
fn function_empty_output_default_value_assignment() {
    // GIVEN
    let result = codegen(
        "
		FUNCTION foo : DINT
		VAR_INPUT
			input1 : DINT;
		END_VAR
		VAR_OUTPUT
			output1 : DINT := 3;
		END_VAR
		VAR_IN_OUT
			inout1 : DINT;
		END_VAR
		END_FUNCTION

		PROGRAM main
		VAR
			var1, var2, var3 : DINT;
		END_VAR
			foo(input1 := var1, output1 => , inout1 := var3);
		END_PROGRAM
		",
    );
    // THEN
    insta::assert_snapshot!(result);
}

#[test]
fn function_empty_inout_assignment() {
    // GIVEN
    let result = codegen(
        "
		FUNCTION foo : DINT
		VAR_INPUT
			input1 : DINT;
		END_VAR
		VAR_OUTPUT
			output1 : DINT;
		END_VAR
		VAR_IN_OUT
			inout1 : DINT;
		END_VAR
		END_FUNCTION

		PROGRAM main
		VAR
			var1, var2, var3 : DINT;
		END_VAR
			foo(input1 := var1, output1 => var2, inout1 := );
		END_PROGRAM
		",
    );
    // THEN
    insta::assert_snapshot!(result);
}

#[test]
fn function_missing_input_assignment() {
    // GIVEN
    let result = codegen(
        "
		FUNCTION foo : DINT
		VAR_INPUT
			input1 : DINT;
		END_VAR
		VAR_OUTPUT
			output1 : DINT;
		END_VAR
		VAR_IN_OUT
			inout1 : DINT;
		END_VAR
		END_FUNCTION

		PROGRAM main
		VAR
			var1, var2, var3 : DINT;
		END_VAR
			foo(output1 => var2, inout1 := var3);
		END_PROGRAM
		",
    );
    // THEN
    insta::assert_snapshot!(result);
}

#[test]
fn function_missing_input_default_value_assignment() {
    // GIVEN
    let result = codegen(
        "
		FUNCTION foo : DINT
		VAR_INPUT
			input1 : DINT := 10;
		END_VAR
		VAR_OUTPUT
			output1 : DINT;
		END_VAR
		VAR_IN_OUT
			inout1 : DINT;
		END_VAR
		END_FUNCTION

		PROGRAM main
		VAR
			var1, var2, var3 : DINT;
		END_VAR
			foo(output1 => var2, inout1 := var3);
		END_PROGRAM
		",
    );
    // THEN
    insta::assert_snapshot!(result);
}

#[test]
fn function_missing_output_assignment() {
    // GIVEN
    let result = codegen(
        "
		FUNCTION foo : DINT
		VAR_INPUT
			input1 : DINT;
		END_VAR
		VAR_OUTPUT
			output1 : DINT;
		END_VAR
		VAR_IN_OUT
			inout1 : DINT;
		END_VAR
		END_FUNCTION

		PROGRAM main
		VAR
			var1, var2, var3 : DINT;
		END_VAR
			foo(input1 := var1, inout1 := var3);
		END_PROGRAM
		",
    );
    // THEN
    insta::assert_snapshot!(result);
}

#[test]
fn function_missing_output_default_value_assignment() {
    // GIVEN
    let result = codegen(
        "
		FUNCTION foo : DINT
		VAR_INPUT
			input1 : DINT;
		END_VAR
		VAR_OUTPUT
			output1 : DINT := 3;
		END_VAR
		VAR_IN_OUT
			inout1 : DINT;
		END_VAR
		END_FUNCTION

		PROGRAM main
		VAR
			var1, var2, var3 : DINT;
		END_VAR
			foo(input1 := var1, inout1 := var3);
		END_PROGRAM
		",
    );
    // THEN
    insta::assert_snapshot!(result);
}

#[test]
fn function_missing_inout_assignment() {
    // GIVEN
    let result = codegen(
        "
		FUNCTION foo : DINT
		VAR_INPUT
			input1 : DINT;
		END_VAR
		VAR_OUTPUT
			output1 : DINT;
		END_VAR
		VAR_IN_OUT
			inout1 : DINT;
		END_VAR
		END_FUNCTION

		PROGRAM main
		VAR
			var1, var2, var3 : DINT;
		END_VAR
			foo(input1 := var1, output1 => var2);
		END_PROGRAM
		",
    );
    // THEN
    insta::assert_snapshot!(result);
}

#[test]
fn function_default_value_parameter_type() {
    // GIVEN
    let result = codegen(
        "
		TYPE myType : DINT := 20; END_TYPE

		FUNCTION foo : DINT
		VAR_INPUT
			input1 : myType;
		END_VAR
		VAR_OUTPUT
			output1 : myType;
			output2 : myType;
		END_VAR
		VAR_IN_OUT
			inout1 : DINT;
		END_VAR
		END_FUNCTION

		PROGRAM main
		VAR
			var1, var2, var3 : DINT;
		END_VAR
			foo(output2 => , inout1 := var3);
		END_PROGRAM
		",
    );
    // THEN
    insta::assert_snapshot!(result);
}

#[test]
fn program_all_parameters_assigned_explicit() {
    // GIVEN
    let result = codegen(
        "
		PROGRAM prog
		VAR_INPUT
			input1 : DINT;
		END_VAR
		VAR_OUTPUT
			output1 : DINT;
		END_VAR
		VAR_IN_OUT
			inout1 : DINT;
		END_VAR
		END_PROGRAM

		PROGRAM main
		VAR
			var1, var2, var3 : DINT;
		END_VAR
			prog(input1 := var1, output1 => var2, inout1 := var3);
			prog(inout1 := var3, input1 := var1, output1 => var2);
		END_PROGRAM
		",
    );
    // THEN
    insta::assert_snapshot!(result);
}

#[test]
fn program_all_parameters_assigned_implicit() {
    // GIVEN
    let result = codegen(
        "
		PROGRAM prog
		VAR_INPUT
			input1 : DINT;
		END_VAR
		VAR_OUTPUT
			output1 : DINT;
		END_VAR
		VAR_IN_OUT
			inout1 : DINT;
		END_VAR
		END_PROGRAM

		PROGRAM main
		VAR
			var1, var2, var3 : DINT;
		END_VAR
			prog(var1, var2, var3);
		END_PROGRAM
		",
    );
    // THEN
    insta::assert_snapshot!(result);
}

#[test]
fn program_empty_inout_assignment() {
    // GIVEN
    let result = codegen(
        "
		PROGRAM prog
		VAR_INPUT
			input1 : DINT;
		END_VAR
		VAR_OUTPUT
			output1 : DINT;
		END_VAR
		VAR_IN_OUT
			inout1 : DINT;
		END_VAR
		END_PROGRAM

		PROGRAM main
		VAR
			var1, var2, var3 : DINT;
		END_VAR
			prog(input1 := var1, output1 => var2, inout1 := );
		END_PROGRAM
		",
    );
    // THEN
    insta::assert_snapshot!(result);
}

#[test]
fn program_missing_input_assignment() {
    // GIVEN
    let result = codegen(
        "
		PROGRAM prog
		VAR_INPUT
			input1 : DINT;
		END_VAR
		VAR_OUTPUT
			output1 : DINT;
		END_VAR
		VAR_IN_OUT
			inout1 : DINT;
		END_VAR
		END_PROGRAM

		PROGRAM main
		VAR
			var1, var2, var3 : DINT;
		END_VAR
			prog(output1 => var2, inout1 := var3);
		END_PROGRAM
		",
    );
    // THEN
    insta::assert_snapshot!(result);
}

#[test]
fn program_missing_output_assignment() {
    // GIVEN
    let result = codegen(
        "
		PROGRAM prog
		VAR_INPUT
			input1 : DINT;
		END_VAR
		VAR_OUTPUT
			output1 : DINT;
		END_VAR
		VAR_IN_OUT
			inout1 : DINT;
		END_VAR
		END_PROGRAM

		PROGRAM main
		VAR
			var1, var2, var3 : DINT;
		END_VAR
			prog(input1 := var1, inout1 := var3);
		END_PROGRAM
		",
    );
    // THEN
    insta::assert_snapshot!(result);
}

#[test]
fn program_accepts_empty_statement_as_input_param() {
    // GIVEN
    let result = codegen(
        "
		PROGRAM prog
		VAR_INPUT
			in1: DINT;
			in2: DINT;
		END_VAR
		END_PROGRAM

		PROGRAM main
			prog(in1 := 1, in2 := );
		END_PROGRAM
		",
    );

    // THEN
    insta::assert_snapshot!(result);
}

#[test]
fn program_accepts_empty_statement_as_output_param() {
    // GIVEN
    let result = codegen(
        "
		PROGRAM prog
		VAR_OUTPUT
			out1 : DINT;
			out2 : DINT;
		END_VAR
			out1 := 1;
			out2 := 2;
		END_PROGRAM

		PROGRAM main
		VAR
			x : DINT;
		END_VAR
			prog( out1 => x, out2 => );
		END_PROGRAM
		",
    );

    // THEN
    insta::assert_snapshot!(result);
}

#[test]
fn fb_accepts_empty_statement_as_input_param() {
    // GIVEN
    let result = codegen(
        "
		FUNCTION_BLOCK fb_t
		VAR_INPUT
			in1: DINT;
			in2: DINT;
		END_VAR
		END_FUNCTION_BLOCK

		PROGRAM main
		VAR
			fb : fb_t;
		END_VAR
			fb(in1 := 1, in2 := );
		END_PROGRAM
		",
    );

    // THEN
    insta::assert_snapshot!(result);
}

#[test]
fn fb_accepts_empty_statement_as_output_param() {
    // GIVEN
    let result = codegen(
        "
		FUNCTION_BLOCK fb_t
		VAR_OUTPUT
			out1 : DINT;
			out2 : DINT;
		END_VAR
		END_FUNCTION_BLOCK

		PROGRAM main
		VAR
			fb : fb_t;
			x : DINT;
		END_VAR
			fb( out1 => x, out2 => );
		END_PROGRAM
		",
    );

    // THEN
    insta::assert_snapshot!(result);
}

#[test]
fn function_accepts_empty_statement_as_input_param() {
    // GIVEN
    let result = codegen(
        "
		FUNCTION foo
		VAR_INPUT
			in1: DINT;
			in2: DINT;
		END_VAR
		END_FUNCTION

		PROGRAM main
			foo(in1 := 1, in2 := );
		END_PROGRAM
		",
    );

    // THEN
    insta::assert_snapshot!(result);
}

#[test]
fn function_accepts_empty_statement_as_output_param() {
    // GIVEN
    let result = codegen(
        "
		FUNCTION foo
		VAR_OUTPUT
			out1 : DINT;
			out2 : DINT;
		END_VAR
		END_FUNCTION

		PROGRAM main
		VAR
			x: DINT;
		END_VAR
			foo( out1 => x, out2 => );
		END_PROGRAM
		",
    );

    // THEN
    insta::assert_snapshot!(result);
}
