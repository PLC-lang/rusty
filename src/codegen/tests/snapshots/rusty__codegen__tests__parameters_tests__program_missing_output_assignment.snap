---
source: src/codegen/tests/parameters_tests.rs
expression: result
snapshot_kind: text
---
; ModuleID = '<internal>'
source_filename = "<internal>"
target datalayout = "e-m:e-p270:32:32-p271:32:32-p272:64:64-i64:64-f80:128-n8:16:32:64-S128"
target triple = "x86_64-pc-linux-gnu"

%prog = type { i32, i32, i32* }
%main = type { i32, i32, i32 }

@prog_instance = global %prog zeroinitializer
@main_instance = global %main zeroinitializer

define void @prog(%prog* %0) {
entry:
  %input1 = getelementptr inbounds %prog, %prog* %0, i32 0, i32 0
  %output1 = getelementptr inbounds %prog, %prog* %0, i32 0, i32 1
  %inout1 = getelementptr inbounds %prog, %prog* %0, i32 0, i32 2
  ret void
}

define void @main(%main* %0) {
entry:
  %var1 = getelementptr inbounds %main, %main* %0, i32 0, i32 0
  %var2 = getelementptr inbounds %main, %main* %0, i32 0, i32 1
  %var3 = getelementptr inbounds %main, %main* %0, i32 0, i32 2
  %load_var1 = load i32, i32* %var1, align 4
  store i32 %load_var1, i32* getelementptr inbounds (%prog, %prog* @prog_instance, i32 0, i32 0), align 4
  store i32* %var3, i32** getelementptr inbounds (%prog, %prog* @prog_instance, i32 0, i32 2), align 8
  call void @prog(%prog* @prog_instance)
  ret void
}
